pallets/xcm-asset-manager/src/lib.rs

```diff
@@ -0,0 +1,423 @@
+			AssetTypeId::<T>::get(asset_type)
+		}
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The Asset Id. This will be used to register the asset in Assets
+		type AssetId: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Ord
+			+ PartialOrd
+			+ Copy
+			+ HasCompact
+			+ MaxEncodedLen;
+
+		/// The Foreign Asset Kind.
+		type ForeignAssetType: Parameter + Member + Ord + PartialOrd;
+
+		/// The units in which we record balances.
+		type Balance: Member + Parameter + AtLeast32BitUnsigned + Default + Copy + MaxEncodedLen;
+
+		/// Origin that is allowed to create and modify asset information for foreign assets
+		type ForeignAssetModifierOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The currency mechanism in which we reserve deposits for local assets.
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		type WeightInfo: WeightInfo;
+	}
+
+	/// An error that can occur while executing the mapping pallet's logic.
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetAlreadyExists,
+		AssetTypeDoesNotExist,
+		AssetIdDoesNotExist,
+		DefaultAssetTypeRemoved,
+		AssetIdLimitReached,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The foreign asset updated.
+		ForeignAssetMetadataUpdated { asset_id: T::AssetId, metadata: AssetMetadata<BalanceOf<T>> },
+		/// AssetTracker manipulated
+		ForeignAssetTrackerUpdated { old_asset_tracker: T::AssetId, new_asset_tracker: T::AssetId },
+		/// New asset with the asset manager is registered
+		ForeignAssetTypeRegistered { asset_id: T::AssetId, asset_type: T::ForeignAssetType },
+		/// New Event gives the info about involved asset_id, removed asset_type, and the new
+		/// default asset_id and asset_type mapping after removal
+		ForeignAssetTypeRemoved {
+			asset_id: T::AssetId,
+			removed_asset_type: T::ForeignAssetType,
+			default_asset_type: T::ForeignAssetType,
+		},
+		/// Changed the amount of units we
+		/// are charging per execution second for a given asset
+		UnitsPerSecondChanged { asset_id: T::AssetId, units_per_second: u128 },
+	}
+
+	/// Mapping from an asset id to asset type.
+	/// This is mostly used when receiving transaction specifying an asset directly,
+	/// like transferring an asset from this chain to another.
+	#[pallet::storage]
+	#[pallet::getter(fn asset_id_type)]
+	pub type AssetIdType<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, T::ForeignAssetType, OptionQuery>;
+
+	/// Reverse mapping of AssetIdType. Mapping from an asset type to an asset id.
+	/// This is mostly used when receiving a multilocation XCM message to retrieve
+	/// the corresponding asset in which tokens should me minted.
+	#[pallet::storage]
+	#[pallet::getter(fn asset_type_id)]
+	pub type AssetTypeId<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::ForeignAssetType, T::AssetId, OptionQuery>;
```

Comment:
IMO it's not best practice to use 2 maps for this data structure. As we don't have a native bidirectional map container, if Id<>Type is 1<>N, we could use a map of <id, vector<Type>>. A prerequisite of this is, we don't have too many registered ids, otherwise, the lookup of a type would be slow.
The advantage of this solution is less complicated, as the current one is bit too complicated and redundant. We can discuss about this later.

Reply:
One of advantage by your modification is, if using vector, we can query the existing asset_type of specific asset_id by storage without RPC. 

Disadvantage is the weight of transaction will be larger.

Reply:
https://docs.substrate.io/v3/advanced/hash-collections/
https://stackoverflow.com/questions/34305901/advantage-of-binary-search-tree-over-vector-in-c#:~:text=BST%20and%20other%20tree%20structures,memory%20usage%20while%20re%2Dsizing.

Reply:
todo: better wrapper of internal function when modify asset_id and asset_type.